Revealing pre-images beyond the end of the enrollment cycle

hewison-chris opened the pull request:
Now that the pre-images are based on height we can reveal them into the next cycle and thus reduce the risk of being slashed when the next cycle starts.

hewison-chris committed:
Trivial removal of unused function `getAvailableHeight`

hewison-chris committed:
Allow preimage reveal beyond enrollemnt

If the pre-images are only revealed up to the end of the enrollment
cycle then there is more risk for the validator to get slashed in the
first block of the new cycle.

hewison-chris committed:
Do not update stored preimage if we have higher

hewison-chris committed:
Do not reset validator enroll_key

We can still use the stored `enroll_key` to send valid preimages if the
timer fires before we re-enroll. This will improve preimage revealing as
we will still share it with other nodes whilst in the process of
re-enrolling.

hewison-chris committed:
Trivial removal of unused function `resetNodeEnrollment`

hewison-chris committed:
Ignore envelopes that originate from self

We can return early once we know that this envelope originates from this
node. Also we avoid `SCP indicated invalid envelope` in logs which is
only because SCP detects it is from current node.

hewison-chris committed:
Trivial logging of function name in handleSCPEnvelope

hewison-chris committed:
Ignore envelope for height after next block

Also trigger the catchup task as we are most likely behind the other
nodes in block height.

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -475,22 +475,19 @@ public class EnrollmentManager
+        in Height expected_height)
         @safe
     {
-        const enrolled = this.validator_set.getEnrolledHeight(height, this.enroll_key);
-        if (enrolled == ulong.max)
-            return false;
-
-        assert(height >= enrolled);
```
I don't think these lines should be removed.

hewison-chris replied:
I think it is better to reveal our pre-image even if we have not yet read the enrollment during `acceptBlock`. This way we are less likely to get slashed if other nodes are already nominating the next block but we are a bit behind.

linked0 replied:
Isn't there any risk of revealing pre-images even if the node is not a validator? I was being worried about that. I will look into it more.

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -475,22 +475,19 @@ public class EnrollmentManager
 
-    public bool getNextPreimage (out PreImageInfo preimage, in Height height)
+    public bool getNextPreimage (out PreImageInfo preimage, in Height height,
+        in Height expected_height)
```
I think it's better to leave as it was because it's right for the ledger to determine which height of pre-image should be revealed.

hewison-chris replied:
The problem is that the ledger does not have access to the `clock`. Do you suggest passing the clock to the `Ledger`?

linked0 replied:
Ah, I'm very sorry. I meant that a validator should determine which height of pre-image should be revealed because the code looks complicated in `getNextPreimage`.

linked0 replied:
I'm sorry that the above comment is not correct. I meant that validators should determine which height of pre-image should be revealed on `onPreImageRevealTimer` because the code of getNextPreimage looks complicated and it's not the EM's role.

hewison-chris replied:
I will move the logic up into `Validator`.

hewison-chris replied:
Updated

linked0 replied:
I meant the function should be as it was. The change is not related to what this PR is to achieve, I think. I'm not sure this change is really what we need for the moment.

hewison-chris replied:
This change is to help prevent slashing due to late blocks so seems related to me.

hewison-chris committed:
Trivial: include file and line of test in assert

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -138,11 +141,13 @@ public class ValidatorSet
-                "VALUES (?, ?, ?)",
-                enroll.utxo_key, height.value,
-                enroll.commitment);
+            // Only add or update if we do not have a newer preimage already
+            if (this.getPreimage(enroll.utxo_key).height <= height)
+                this.db.execute("INSERT OR REPLACE INTO preimages " ~
+                    "(key, height, preimage) " ~
+                    "VALUES (?, ?, ?)",
+                    enroll.utxo_key, height.value,
+                    enroll.commitment);
```
Wouldn't it be better to just call `addPreimage` here (after the next line though) ?

hewison-chris replied:
In `addPreimage` it always does an update and not an insert. Do you mean update that function to handle both cases?

hewison-chris replied:
Actually that is better as we can check the height for any pre-image added is more than stored height.

Geod24 replied:
Hum right, I missed that. I'm just looking at the statement and wondering if we could improve it be removing the conditional. But I haven't found a good approach, so let's go ahead with the current state.

hewison-chris replied:
When we receive a `pre-mage` via a post in `FullNode` we don't currently check the height is higher so we need the check anyway.

hewison-chris replied:
Just realized we already check in the `addPreimage` function that the height is higher than previous.

linked0 commented on source/agora/test/ValidatorRecurringEnrollment.d:
```diff
@@ -131,12 +131,19 @@ unittest
+    assert(block20.header.enrollments.length == 3);
+
+    // This block has only half of the validators enrolled
+    network.generateBlocks(iota(3), Height(GenesisValidatorCycle + 1), true);
```
Shouldn't all the nodes be able to generate blocks regardless of the number of validators?

hewison-chris replied:
In this test only the first 3 enrollments were put into the block so only 3 validators will be active in that block. The other 3 will fetch the block during catchup. The `generateBlocks` calls `addBlock` which assumes enrollments are in the last block of the cycle. In this test that is not the case for the last 3 nodes.

linked0 replied:
I see what you mean. As far as I know, the last half of validators who are not re-enrolled yet can not have the pre-images of current validators at the height of 20.

linked0 reviewed (commented):
Good.
It would be better to separate into more atomic PRs, like a PR for trivial change and a PR for revealing pre-images if there are so many commits in a PR.

hewison-chris commented:
> Good. It would be better to separate into more atomic PRs, like a PR for trivial change and a PR for revealing pre-images if there are so many commits in a PR.

I agree but whilst we do not have green builds it is difficult to get those trivial changes approved. I will split this once I have green again. Until then I need the changes to help with the debugging.

linked0 commented on source/agora/test/Base.d:
```diff
@@ -776,29 +775,28 @@ public class TestAPIManager
+            foreach (enroll; enrolls)
+                retryFor(node.client.getPreimages(Set!Hash.from(enroll.utxo_key.only))
+                    .any!(preimage => preimage.height >= height),
+                        timeout, format!"Client #%s (%s) has no preimage at height %s for node enrolled with utxo %s"
```
```suggestion
                        timeout, format!"Client #%s (%s) has no preimage for height %s for node enrolled with utxo %s"
```
I suggest it because it looks like the current height. But you could determine better than me for that.

hewison-chris replied:
I agree will change.

hewison-chris replied:
Updated

hewison-chris committed:
Trivial logging of peer identities in `emitEnvelope`

hewison-chris committed:
Only re-arm task timers if not pending

linked0 reviewed (approved):
LGTM, the last three commits could be another PR tho.

hewison-chris committed:
Trivial addition of address of failed node in test

hewison-chris committed:
Trivial log improvement for emitEnvelope

hewison-chris committed:
Trivial debug log for computeTimeout

hewison-chris committed:
Trivial addition of function to clear all logs

hewison-chris committed:
Test code: using enrollments from test case

In some tests not all the nodes will have pre-images at a certain
height. In these tests, rather than relying on the block header at the
end of the enrollment cycle, we can pass the enrollments of those nodes
that should have a preimage at that height.

hewison-chris committed:
Use round timeout once nominating

When nomination is started then use the computed round timeout to also
trigger the next nomination attempt.

hewison-chris committed:
Send envleopes again if not new nomination

This will cause a large volume of envelope gossiping so will need to be
optimized to only happen when we are not making progress. But let's see
if this actually helps in ci first.

hewison-chris committed:
Integration test: Reduce nomination interval

To prevent lots of empty blocks with a block interval of 5 seconds we
need to check if it is time to nominate more frequently.

hewison-chris committed:
Correct the log for rejected pre-image

It was writing the pre-image hash where it should be the utxo

hewison-chris committed:
Check that utxo is valid in pre-image

Only send pre-image if the utxo is set.

hewison-chris commented:
Closing as it has been split up.
